service/history/historyReplicator.go

```diff
@@ -983,22 +997,28 @@ func (r *historyReplicator) replicateWorkflowStarted(ctx context.Context, contex
 }
 
 func (r *historyReplicator) flushCurrentWorkflowBuffer(ctx context.Context, domainID string, workflowID string,
-	logger bark.Logger) (string, int64, bool, error) {
+	logger bark.Logger) (runID string, nextEventID int64, createTaskID int64, isRunning bool, retError error) {
 	currentContext, currentMutableState, currentRelease, err := r.getCurrentWorkflowMutableState(ctx, domainID,
 		workflowID)
 	if err != nil {
-		return "", 0, false, err
+		return "", 0, 0, false, err
 	}
+	defer func() { currentRelease(retError) }()
 	// since this new workflow cannot make progress due to existing workflow being open
 	// try flush the existing workflow's buffer see if we can make it move forward
 	// First check if there are events which needs to be flushed before applying the update
 	err = r.flushReplicationBuffer(ctx, currentContext, currentMutableState, logger)
-	currentRelease(err)
 	if err != nil {
 		logError(logger, "Fail to flush buffer for current workflow.", err)
-		return "", 0, false, err
+		return "", 0, 0, false, err
 	}
-	return currentContext.getExecution().GetRunId(), currentMutableState.GetNextEventID(), currentMutableState.IsWorkflowExecutionRunning(), nil
+	runID = currentContext.getExecution().GetRunId()
+	nextEventID = currentMutableState.GetNextEventID()
+	createTaskID = currentMutableState.GetExecutionInfo().CreateTaskID
+	isRunning = currentMutableState.IsWorkflowExecutionRunning()
+	currentRelease(err)
```

Comment:
You already have a release in a defer function.  Is that really needed?

Reply:
last time there were panic within the lock, using defer right after getting the lock, will at least make the deadlock go away